[pipelining] Add pipeline schedules

kwen2501 opened the pull request:
1. Add pipeline schedules:
- GPipe
- 1F1B
- Interleaved 1F1B
- LoopedBFS

2. Add basic forward and backward tests: 
test_schedule.py

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #125975
* #125729



cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225 @chauhang @d4l3k

kwen2501 committed:
[pipelining] Add pipeline schedules

[ghstack-poisoned]

wconstab reviewed (approved):
lgtm, again assuming just code motion i am not reviewing carefully

nit: the ScheduleLoopedBFS was missing from PR desc but i added it

kwen2501 committed:
Update on "[pipelining] Add pipeline schedules"

1. Add pipeline schedules:
- GPipe
- 1F1B
- Interleaved 1F1B
- LoopedBFS

2. Add basic forward and backward tests: 
test_schedule.py




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge

kwen2501 committed:
Update on "[pipelining] Add pipeline schedules"

1. Add pipeline schedules:
- GPipe
- 1F1B
- Interleaved 1F1B
- LoopedBFS

2. Add basic forward and backward tests: 
test_schedule.py




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 committed:
Update on "[pipelining] Add pipeline schedules"

1. Add pipeline schedules:
- GPipe
- 1F1B
- Interleaved 1F1B
- LoopedBFS

2. Add basic forward and backward tests: 
test_schedule.py




cc mrshenli pritamdamania87 zhaojuanmao satgera gqchen aazzolini osalpekar jiayisuse H-Huang awgu penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225 chauhang d4l3k

[ghstack-poisoned]

kwen2501 commented:
@pytorchbot merge
